src/sdk/namespace_fs.js

```diff
@@ -1964,69 +1974,81 @@ class NamespaceFS {
-                if (!restore_expiry) {
-                    dbg.log0('namespace_fs.restore_object: state - (!restore_request, restore_ongoing, !restore_expiry)');
-                    throw new S3Error(S3Error.RestoreAlreadyInProgress);
-                }
-            } else {
-                const now = new Date();
+            // 2 states
+            if (
+                (restore_request && !restore_expiry) ||
+                (restore_request && restore_expiry <= now)
+            ) {
+                dbg.log0(
+                    'namespace_fs.restore_object: state - (restore_request, !restore_expiry)',
+                    'or',
+                    'state - (restore_request, restore_expiry <= now)'
+                );
 
-                // The only entity that can remove the restore request is the underlying
-                // restore mechanism, implying that at some point it must have picked
-                // up the request to restore and finished the process.
-                if (restore_expiry && restore_expiry > now) {
-                    dbg.log0('namespace_fs.restore_object: state - (!restore_request, !restore_ongoing, restore_expiry > now)');
+                throw new S3Error(S3Error.RestoreAlreadyInProgress);
+            }
 
-                    const expires_on = now;
-                    expires_on.setDate(expires_on.getDate() + params.days);
+            // 1 state
+            if (restore_request && restore_expiry > now) {
+                // Invalid state, should have been removed
+                // by the backend - Might be an unfortunate
+                // race where we read the stat just before
+                // removing the request attr and after adding
+                // the expiry
+                dbg.error('namespace_fs.restore_object: state - (restore_request, restore_expiry > now)');
+                throw new RpcError('INTERNAL_ERROR');
+            }
 
-                    await file.replacexattr(fs_context, {
-                        [GlacierBackend.XATTR_RESTORE_EXPIRY]: expires_on.toISOString(),
-                    });
+            // 2 states
+            if (
+                (!restore_request && !restore_expiry) ||
+                (!restore_request && restore_expiry <= now)
+            ) {
+                dbg.log0(
+                    'namespace_fs.restore_object: state - (!restore_request, !restore_expiry)',
+                    'or',
+                    '(restore_request, restore_expiry <= now)'
+                );
+
+                // First add it to the log and then add the extended attribute as if we fail after
+                // this point then the restore request can be triggered again without issue but
+                // the reverse doesn't works.
+                await this.append_to_restore_wal(file_path);
+
+                await file.replacexattr(fs_context, {
+                    [GlacierBackend.XATTR_RESTORE_REQUEST]: params.days.toString(),
```

Comment:
Pehaps better to use restore_request value here which was already parsed and checked for validity instead of param.days

Reply:
Yes, in s3_post_object_restore.js